### PR TITLE
[master] fix #1950: Duplicate objects in UnitOfWorkImpl.primaryKeyToNewObjects

### LINKED ISSUE #1950: Duplicate objects in UnitOfWorkImpl.primaryKeyToNewObjects
When `UnitOfWorkImpl#registerNotRegisteredNewObjectForPersist` is called, the object is potentially added twice to the primaryKeyToNewObjects map.
First when the `assignSequenceNumber` method is called (assuming a sequence is used) and a second time when calling `registerNewObjectClone` (which calls `addNewObjectToPrimaryKeyToNewObjects`).

When the object is later unregistered, it is only removed once from the map (see `removeObjectFromPrimaryKeyToNewObjects`).

Subsequent calls on `EntityManger.find` will still find the remaining object, which should not be the case.

The following code snippet illustrates the behaviour:

```java
  @Test
  public void test() {
    Entity entity = new Entity();
    this.entityManager.persist(entity);
    
    Entity savedEntity = this.entityManager.find(Entity.class, entity.getId());
    this.entityManager.remove(savedEntity);
    
    savedEntity = this.entityManager.find(Entity.class, entity.getId());
    assertThat(savedEntity, nullValue());
  }
```

The assertion fails, because the object is still in UnitOfWorkImpl.primaryKeyToNewObjects as it was added twice but only removed once.

Version affected: 2.7.13
Last version with expected behaviour: 2.7.12
Change introduced with https://github.com/eclipse-ee4j/eclipselink/pull/1843

### PR DESCRIPTION
**Pull request by Zuplyx**
- Fixes #1950 

~Objects are now no longer added to `primaryKeyToNewObjects` in `assignSequenceNumber`.~
I've refactored the code to use an `IdentityHashSet` instead of `ArrayList` for `primaryKeyToNewObjects` to prevent duplicates in the future.
Additionally I added two UnitTests. One verifies that calling `registerNewObjectForPersist` does not result in duplicates in `primaryKeyToNewObjects`. The other tests the complete `EntityManager` persist, remove, find workflow from the original issue.

### CODE REVIEW

**Comment by rfelcman** [comment]
Please rebase this PR against latest master. There is fix for the GitHub Actions #1952 .

**Comment by Zuplyx** [comment]
I made a false assumption for my first commit: We still need to add objects to `primaryKeyToNewObjects` in `assignSequenceNumber`, as doing otherwise breaks the use case where a primary key is later assigned to an already registered object. 
Therefore, the solution to prevent duplicates is to simply switch to an `IdentityHashSet`.


@rfelcman Should I create the PR for 2.7 now or would you like to review this one first?
Also I do not know why the ECA check fails for this PR. If I manually verify the email, it says I have a valid ECA.

**Review by rfelcman — APPROVED** [review]
LGTM except copyright comment

File: foundation/eclipselink.core.test/src/it/java/org/eclipse/persistence/testing/tests/unitofwork/UOWPrimaryKeyToNewObjectsDuplicateObjectsTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 1998, 2023 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Zuplyx) [reply foundation/eclipselink.core.test/src/it/java/org/eclipse/persistence/testing/tests/unitofwork/UOWPrimaryKeyToNewObjectsDuplicateObjectsTest.java]: Fixed with the new commit

**Pull request merged**